### ISSUE TITLE
Attribute table layout doesn't update

### ISSUE THREAD

**Issue by Cucumerlu** [qgis/QGIS#43058]
I have selected some columns from attribute table in print composer linked to layer A but when I open my layout in another project with that same layer A, it doesn't refresh/update.

![fig_1](https://user-images.githubusercontent.com/78080818/116851489-b6aa9500-abf2-11eb-9ea5-b14740e9c5e6.png)
![fig_2](https://user-images.githubusercontent.com/78080818/116851528-cc1fbf00-abf2-11eb-80be-7c23c1ff1c67.png)

**Comment by gioman** [comment]
@Cucumerlu 

Please:

add QGIS version is operating system

add detailed steps to replicate

add a sample project with data

**Comment by Cucumerlu** [comment]
@gioman I use QGIS ver 3.18.1 but the problem is the same with ver 3.16.
The steps to replicate the process are:
- open a layout composer
- create a attribute table layout (with layer A) --> select (from object properties, attribute) some columns
- save layout model
- open new project with the same layer A
- open the saved model layout
at this point I expected to see the same attribute table layout with the same column selection instead I see the image shown above.
I haven't a project to show...I works with customer data that I can't publish.

**Comment by gioman** [comment]
>     * works with customer data t

@Cucumerlu does it happens just with that data or with any data?

**Comment by Cucumerlu** [comment]
@gioman with any data

**Comment by gioman** [comment]
> * open a layout composer
> 
> * create a attribute table layout (with layer A) --> select (from object properties, attribute) some columns
> 
> * save layout model
> 
> * open new project with the same layer A
> 
> * open the saved model layout
>   at this point I expected to see the same attribute table layout with the same column selection instead I see the image shown above.

@Cucumerlu just tested on QGIS master, and works fine here.

Can you add a sample project (with any data) and the template you saved? Thanks.

**Comment by Cucumerlu** [comment]
I attached a point layer and a qpt file
[Bug_test.zip](https://github.com/qgis/QGIS/files/6415547/Bug_test.zip)

**Comment by gioman** [comment]
> I attached a point layer and a qpt file

@Cucumerlu in the table in the composer your attributes have names that **do not** correspond to the attributes in the layer you attached. Have you considered that this may be the cause of the problem you are seeing? in fact If I don't use the template you attached, but I create a new one and re-use it in a new project it works as expected.

**Comment by Cucumerlu** [comment]
Sorry, I wrong layer...attribute table is related to the line layer (attached below). The problem is the same, I think.
[layer_line.zip](https://github.com/qgis/QGIS/files/6417084/layer_line.zip)

**Comment by gioman** [comment]
> Sorry, wrong layer...attribute table is related to the line layer (attached below). The problem is the same, I think.

@Cucumerlu same thing as the other layer, no match between the layer and the layout.

**Comment by Cucumerlu** [comment]
@gioman excuse me but I don't understand: I attached 2 images with table layout and attribute table...the fields used are the same.
![attribute_layout fields](https://user-images.githubusercontent.com/78080818/116969114-04d49c80-acb6-11eb-800b-c07f981cd246.PNG)
![attribute_table_fields](https://user-images.githubusercontent.com/78080818/116969115-056d3300-acb6-11eb-89e6-1c1c1401d2d2.PNG)

**Comment by gioman** [comment]
> @gioman excuse me but I don't understand:

@Cucumerlu the field names in the expressions are Uppercase, the fields in your table are Lowercase, so they are different.

**Comment by Cucumerlu** [comment]
@gioman I think that it is not case sensitive BUT I
![attribute_layout fields_lowercase](https://user-images.githubusercontent.com/78080818/116985425-bbdc1280-accc-11eb-89c1-39ff179a2a86.PNG)
![no_update](https://user-images.githubusercontent.com/78080818/116985427-bc74a900-accc-11eb-8b2e-929b494dca61.PNG)

 changed fields name in attribute layout with lowercase but the problem is the same

**Comment by gioman** [comment]
> I changed fields name in attribute layout with lowercase

@Cucumerlu can you attach the new template?

**Comment by Cucumerlu** [comment]
@gioman yes, of corse! The modified layout (with lowercase fields) is related to the table "DATI RELATIVI AGLI ALLACCIAMENTI PRESENTI". Thank you
[Lario_mono_ lowercase.zip](https://github.com/qgis/QGIS/files/6420205/Lario_mono_.lowercase.zip)

**Comment by gioman** [comment]
> @gioman yes, of corse! The modified layout (with lowercase fields) is related to the table "DATI RELATIVI AGLI ALLACCIAMENTI PRESENTI". Thank you

@Cucumerlu works fine here:

![Screenshot_20210504_124229](https://user-images.githubusercontent.com/1951107/116998498-39f0e700-acd6-11eb-81bb-2bbd8843771b.png)

**Comment by Cucumerlu** [comment]
@gioman yes, because I have re-loaded the layer BUT the fields selection and the field expressions are disappered...this is the problem. In the table "DATI RELATIVI AGLI ALLACCIAMENTI PRESENTI" you can see the field selected and the expression.

**Comment by gioman** [comment]
@Cucumerlu it is just enough to force re-select the layer that is supposed to populate the table:

![Peek 2021-05-04 13-27](https://user-images.githubusercontent.com/1951107/117003187-76274600-acdc-11eb-8e6f-4ea02b06d571.gif)

**Comment by Cucumerlu** [comment]
Yes, I know (I told the same concept in the previous post ("I have re-loaded the layer") but in this way you totally loose the fields selection and expression.

**Comment by gioman** [comment]
@Cucumerlu Ok, sorry for the confusion. 

The table do not get associated with the layer... this is clear because upon opening the layout and selecting the table the option "show only feature visible within a map" cannot be enabled. But his seems to only happen if the table has fields defined by expression and when creating a layout from a template.

Sound correct as description?

**Comment by gioman** [comment]
> Sound correct as description?

@Cucumerlu answering myself... it seems correct, but also happens **only** with your very specific case/template/table.

Try the following template where I recreated your table by manually copying the expressions and the headers names, it work as expected.

So it seems there is something very peculiar in the table in your template, which may explain the behavior.

[Lario_mono_ lowercase4.zip](https://github.com/qgis/QGIS/files/6421290/Lario_mono_.lowercase4.zip)
.

**Comment by Cucumerlu** [comment]
@gioman I tried the template but the problem is the same. If I recreate manually the table layout, it works fine but when I use the layout model in another project there is the problem :(

**Comment by gioman** [comment]
> @gioman I tried the template but the problem is the same. If I recreate manually the table layout, it works fine but when I use the layout model in another project there is the problem :( 

@Cucumerlu the template I sent you woks fine here, even on new projects.

**Comment by Cucumerlu** [comment]
@gioman as you see in the video attached there is the same problem. When I select the atlas cover layer the table doesn't refresh. I tested this problem with my colleagues: different QGIS version, different PC...the problem is real and not related to my specific layer or project
https://user-images.githubusercontent.com/78080818/117265083-279cb780-ae54-11eb-8a65-fe140a8ecce5.MP4

**Comment by gioman** [comment]
> @gioman as you see in the video attached there is the same problem. When I select the atlas cover layer the table doesn't refresh. I tested this problem with my colleagues: different QGIS version, different PC...the problem is real and not related to my specific layer or project

@Cucumerlu I edited the title to better reflect the issue, and in what circumstances it happens.

**Comment by Cucumerlu** [comment]
> @Cucumerlu I edited the title to better reflect the issue, and in what circumstances it happens.

Thank you
